### PR TITLE
US663374 : redeployable bundle

### PR DESCRIPTION
**Pull request by RaghuReddy**
## Description
Policy author should be able to build annotated redeployable bundles.

## Links
Link Type   | Link
------      | ------
Rally Issue | [US663374]
Func Spec   | [Func Spec]

## Checklist
- [x] Pull Request Created
- [ ] Code Review Completed
- [ ] Veracode scan results addressed
- [ ] SonarQube scan results addressed
- [ ] TPSR has been submitted (if applicable) 
- [ ] Unit tests have been added
- [ ] Integration/Functional test cases have been written and automated
- [ ] Func Spec has been updated as needed
- [ ] Rally Issue(s) are in an engineering completed state

### CODE REVIEW

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java [hunk 1]
```diff
@@ -91,6 +91,7 @@ private Entity buildEncassEntity(AnnotatedBundle annotatedBundle, Bundle bundle,
         String guid = encass.getGuid();
         String id = encass.getId();
         AnnotatedEntity annotatedEntity = annotatedBundle != null ? annotatedBundle.getAnnotatedEntity() : null;
+        final boolean isRedeployableBundle = annotatedEntity != null && annotatedEntity.isBundle() && annotatedEntity.isRedeployable();
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java:94]: isBUndle check is not required entity present in annotatedBundle is always a bundle annotation entity

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java [hunk 2]
```diff
@@ -114,7 +115,7 @@ private Entity buildEncassEntity(AnnotatedBundle annotatedBundle, Bundle bundle,
         properties.putIfAbsent(PALETTE_FOLDER, DEFAULT_PALETTE_FOLDER_LOCATION);
         buildAndAppendPropertiesElement(properties, document, encassAssertionElement);
         Entity entity = getEntityWithNameMapping(ENCAPSULATED_ASSERTION_TYPE, encassName, id, encassAssertionElement);
-        if (!encass.isReusable() && !isAnnotatedEntity(encass, annotatedEntity)) {
+        if ((!encass.isReusable() && !isAnnotatedEntity(encass, annotatedEntity)) || isRedeployableBundle) {
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java:118]: same as above

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/PolicyEntityBuilder.java [hunk 3]
```diff
@@ -376,6 +376,7 @@ Entity buildPolicyEntity(Policy policy, AnnotatedBundle annotatedBundle, Bundle
         String policyNameWithPath = policy.getPath();
         AnnotatedEntity annotatedPolicyEntity = null;
         AnnotatedEntity annotatedEntity = annotatedBundle != null ? annotatedBundle.getAnnotatedEntity() : null;
+        final boolean isRedeployableBundle = annotatedEntity != null && annotatedEntity.isBundle() && annotatedEntity.isRedeployable();
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/PolicyEntityBuilder.java:379]: same as above

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java (hunk 1 above)
Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java]: done

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/PolicyEntityBuilder.java (hunk 3 above)
Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/PolicyEntityBuilder.java]: done

**Comment by nikhilahuja** [comment]
@RaghuReddy Shouldn't this include Metadata change for `redeployable` flag?

**Comment by RaghuReddy** [comment]
> 
> 
> @RaghuReddy Shouldn't this include Metadata change for `redeployable` flag?



> 
> 
> @RaghuReddy Shouldn't this include Metadata change for `redeployable` flag?

it is already taken care while building metadata.

**Comment by nikhilahuja** [comment]
`redeployable` flag in metadata can be true in 2 cases:
1. If the bundle is annotated with `@redeployable` annotation.
2. If the bundle doesn't contain any reusable entity.

**Comment by RaghuReddy** [comment]
> 
> 
> `redeployable` flag in metadata can be true in 2 cases:
> 
>     1. If the bundle is annotated with `@redeployable` annotation.
> 
>     2. If the bundle doesn't contain any reusable entity.

Done.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java [hunk 4]
```diff
@@ -33,13 +34,15 @@ public BundleMetadata build(final AnnotatedBundle annotatedBundle, final Annotat
         final Encass encass = (Encass) annotatedEntity.getEntity();
         final String bundleName = annotatedBundle.getBundleName();
         final String name = bundleName.substring(0, bundleName.indexOf(projectVersion) - 1);
+        final boolean isPolicyOrEncassReusable = annotatedBundle.getEntities(Policy.class).entrySet().stream().anyMatch(entity -> entity.getValue().isReusable()) ||
+                annotatedBundle.getEntities(Encass.class).entrySet().stream().anyMatch(entity -> entity.getValue().isReusable());
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java:38]: You can mode this logic to a different method. And then bellow, you can do like this:
`builder.reusableAndRedeployable(true, annotatedEntity.isRedeployable() || !bundleContainsReusableEntity());`
This way we avoid iterating the policies and encasses if `@redeployable` is set.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java [hunk 5]
```diff
@@ -33,13 +34,15 @@ public BundleMetadata build(final AnnotatedBundle annotatedBundle, final Annotat
...
 
         BundleMetadata.Builder builder = new BundleMetadata.Builder(encass.getType(), encass.getGuid(), name,
                 projectGroupName, projectVersion);
         builder.description(annotatedEntity.getDescription());
         builder.environmentEntities(getEnvironmentDependenciesMetadata(dependentEntities));
         builder.tags(annotatedEntity.getTags());
-        builder.reusableAndRedeployable(true, annotatedEntity.isRedeployable());
+        builder.reusableAndRedeployable(true, !isPolicyOrEncassReusable || annotatedEntity.isRedeployable());
```
Comment (nikhilahuja) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java:45]: Please add JUnits to cover these 2 cases in metadata.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java (hunk 4 above)
Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java]: done.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java (hunk 5 above)
Comment (RaghuReddy) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleMetadataBuilder.java]: Done

**Pull request merged**